File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.cpp
#include <string>
#include <vector>
using namespace std;

int res = 1;


void dfs(vector<vector<bool>>&arr, int start, int n, vector<vector<bool>>&visited) {
    for (int i = 1; i <= n; i++) {
        if (arr[start][i] && !visited[start][i]) {
            visited[start][i] = true; visited[i][start] = true;
            res++;
            dfs(arr, i, n, visited);
        }
    }
    return;
}

void func(vector<vector<bool>>& arr, int start, int n) {
    vector<vector<bool>>visited(n + 1, vector<bool>(n + 1));
    dfs(arr, start, n, visited);
}


int solution(int n, vector<vector<int>> wires) {
    int answer = 1000;

    vector<vector<bool>>arr(n+1, vector<bool>(n+1));
    vector<vector<bool>>visited(n+1, vector<bool>(n+1));    
    
    for (int i = 0; i < wires.size(); i++) {
        int num1 = wires[i][0]; int num2 = wires[i][1];
        arr[num1][num2] = true;
        arr[num2][num1] = true;
    }

    for (int i = 0; i < wires.size(); i++) {
        int num1 = wires[i][0]; int num2 = wires[i][1];
        arr[num1][num2] = false;
        arr[num2][num1] = false;

        res = 1;
        func(arr, num1, n);
        int result1 = res;

        res = 1;
        func(arr, num2, n);
        int result2 = res;
        int result3 = abs(result1 - result2);

        if (answer > result3) {
            answer = result3;
        }

        arr[num1][num2] = true;
        arr[num2][num1] = true;
    }

    return answer;
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] wires) {
        int answer = Integer.MAX_VALUE;

        for (int i = 0; i < wires.length; i++) {
            // i번째 wire 끊기
            List<List<Integer>> graph = buildGraph(n, wires, i);

            // 한 쪽 트리 노드 개수 탐색
            int count = bfs(1, graph, n);
            int diff = Math.abs(count - (n - count));
            answer = Math.min(answer, diff);
        }
        return answer;
    }

    private List<List<Integer>> buildGraph(int n, int[][] wires, int skip) {
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i <= n; i++) graph.add(new ArrayList<>());

        for (int i = 0; i < wires.length; i++) {
            if (i == skip) continue;
            int a = wires[i][0];
            int b = wires[i][1];
            graph.get(a).add(b);
            graph.get(b).add(a);
        }
        return graph;
    }

    private int bfs(int start, List<List<Integer>> graph, int n) {
        boolean[] visited = new boolean[n+1];
        Queue<Integer> q = new LinkedList<>();
        q.offer(start);
        visited[start] = true;
        int count = 1;

        while (!q.isEmpty()) {
            int cur = q.poll();
            for (int next : graph.get(cur)) {
                if (!visited[next]) {
                    visited[next] = true;
                    q.offer(next);
                    count++;
                }
            }
        }
        return count;
    }
}
